exp/addons/controllers/clusterresourcesetbinding_controller.go

```
@@ -0,0 +1,129 @@
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// +kubebuilder:rbac:groups=addons.cluster.x-k8s.io,resources=*,verbs=get;list;watch;create;update;patch;delete
+
+// ClusterResourceSetBindingReconciler reconciles a ClusterResourceSetBinding object
+type ClusterResourceSetBindingReconciler struct {
+	Client client.Client
+	Log    logr.Logger
+
+	scheme *runtime.Scheme
+}
+
+func (r *ClusterResourceSetBindingReconciler) SetupWithManager(mgr ctrl.Manager, options controller.Options) error {
+	_, err := ctrl.NewControllerManagedBy(mgr).
+		For(&addonsv1.ClusterResourceSetBinding{}).
+		Watches(
+			&source.Kind{Type: &clusterv1.Cluster{}},
+			&handler.EnqueueRequestsFromMapFunc{ToRequests: handler.ToRequestsFunc(r.clusterToClusterResourceSetBinding)},
+		).
+		WithOptions(options).
+		WithEventFilter(predicates.ResourceNotPaused(r.Log)).
+		Build(r)
+	if err != nil {
+		return errors.Wrap(err, "failed setting up with a controller manager")
+	}
+
+	r.scheme = mgr.GetScheme()
+	return nil
+}
+
+func (r *ClusterResourceSetBindingReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, reterr error) {
+	ctx := context.Background()
+
+	// Fetch the ClusterResourceSetBinding instance.
+	crsBinding := &addonsv1.ClusterResourceSetBinding{}
+	if err := r.Client.Get(ctx, req.NamespacedName, crsBinding); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	clusterOwner := getOwnerRef(crsBinding.OwnerReferences)
+
+	cluster := &clusterv1.Cluster{}
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: clusterOwner.Name, Namespace: crsBinding.Namespace}, cluster); err != nil {
+		if apierrors.IsNotFound(err) || !cluster.DeletionTimestamp.IsZero() {
+			return r.reconcileDelete(ctx, crsBinding)
+		}
+	}
```

I think these are the cases you need to consider:

1. Some error other than not-found
1. Not-found error
1. Cluster being deleted

It looks like we're swallowing 1 here, and 3 needs to be outside the `err != nil` block.